### PR TITLE
[REVIEW] Tokenizer with rmm integration

### PR DESCRIPTION
**Pull request by brhodes10**
* Updated tokenizer to use rmm
* Added tokenizer benchmark and unit tests

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by brhodes10** [comment]
rerun tests

**Comment by BartleyR** [comment]
Could we also get either @harrism or @jrhemstad to give this a look as well?

**Comment by brhodes10** [comment]
@efajardo-nv when you get a chance can you help point out why the build is failing on `gpuCI/clx/gpu-test/ubuntu18.04-cuda10.2` ? Looks like it just timed out?

**Comment by efajardo-nv** [comment]
rerun tests

**Comment by efajardo-nv** [comment]
rerun tests

**Comment by efajardo-nv** [comment]
rerun tests

**Comment by brhodes10** [comment]
rerun tests

**Comment by efajardo-nv** [comment]
rerun tests

**Comment by efajardo-nv** [comment]
@brhodes10 since this pr still uses pytorch 1.3.1, there won't be official cuda 10.2 support until we move to 1.5.0. we should probably disable the tokenizer unit test until then. we had to remove/disable all pytorch-dependent unit tests and notebook tests because the cuda10.2 test was taking twice as long (about an hour).

**Comment by efajardo-nv** [comment]
Just noticed that tokenizer tests are being skipped for 10.2 but build still gets stuck on this for 14 minutes. Should take only a few seconds. This is the same issue we ran into that led us to remove pytorch-dependent unit tests.
```
collecting ... 
----------------------------- live log collection ------------------------------
"2020-05-20 21:35:41" [INFO] generated new fontManager (font_manager.py:1337)
```

File: ci/gpu/build.sh [hunk 1]
```diff
@@ -43,11 +43,11 @@ python --version
 conda config --set ssl_verify False
 
 logger "conda install required packages"
-conda install -y -c conda-forge -c rapidsai -c rapidsai-nightly -c nvidia \
+conda install -y -c nvidia -c rapidsai -c rapidsai-nightly -c conda-forge -c defaults \
```
Comment (kkraus14) [inline ci/gpu/build.sh:46]: I think you should be able to remove the channels from this command as they're controlled within CI now.

File: conda/environments/clx_dev.yml [hunk 2]
```diff
@@ -16,6 +16,7 @@ dependencies:
 - pytorch==1.3.1
 - torchvision=0.4.2
 - cython>=0.29,<0.30
+- rmm
```
Comment (kkraus14) [inline conda/environments/clx_dev.yml:19]:
```suggestion
- rmm=0.14.*
```

File: cpp/src/basicTokenizer.cu [hunk 3]
```diff
@@ -255,15 +254,15 @@ std::pair<ptr_length_pair<uint32_t*>, ptr_length_pair<uint32_t*>> GpuBasicTokeni
...
+  std::vector<uint32_t> sentence_offsets(num_offsets);
   uint32_t start_copy = 0;
   for(int i = 0; i < offset_size; ++i){
     sentence_offsets[i] = start_copy;
     start_copy += offsets[i];
   }
   sentence_offsets[offset_size] = start_copy;
 
-  assertCudaSuccess(cudaMemcpy(device_sentence_offsets, sentence_offsets, sizeof(*device_sentence_offsets) * num_offsets, cudaMemcpyHostToDevice));
+  device_sentence_offsets = sentence_offsets;
```
Comment (harrism) [inline cpp/src/basicTokenizer.cu:265]: Delete if you make the change above.

```suggestion
```

File: cpp/src/basicTokenizer.cu [hunk 4]
```diff
@@ -141,7 +141,7 @@ __global__ void gpuBasicTokenizer(const unsigned char* sentences,  uint32_t* dev
 }
 
 
-void transfer_cp_data_to_device(uint32_t **device_cp_metadata, uint64_t **device_aux_data) {
+void transfer_cp_data_to_device(rmm::device_vector<uint32_t>& device_cp_metadata, rmm::device_vector<uint64_t>& device_aux_data) {
```
[the remark this answers is not in the document]
Comment (brhodes10) [reply cpp/src/basicTokenizer.cu]: > Where are `cp_data` and `aux_data` defined? This is a free function, so for this to compile those must be global variables, right?!

I have found these in the `cp_data.h` file. I have separated them out into `cp_data_vec.h` for now to resolve some conflicts with the updates to the basicTokenizer constructor - hope thats ok

File: cpp/src/data_transfer_utils.cuh [hunk 5]
```diff
@@ -14,9 +18,8 @@ static void gpuCheck(cudaError_t err, const char *file, int line) {
 #define assertCudaSuccess(cu_err) {gpuCheck((cu_err), __FILE__, __LINE__);}
 
 template<typename T>
-void malloc_and_copy_vec_to_device(T** dest_ptr, std::vector<T> vec) {
-  assertCudaSuccess(cudaMalloc(dest_ptr, sizeof(T) * vec.size()));
-  assertCudaSuccess(cudaMemcpy(*dest_ptr, vec.data(), vec.size() * sizeof(T),
-                               cudaMemcpyHostToDevice));
+void malloc_and_copy_vec_to_device(rmm::device_vector<T>& dest, std::vector<T> const& vec) {
```
[the remark this answers is not in the document]
Comment (brhodes10) [reply cpp/src/data_transfer_utils.cuh]: Removed function

File: cpp/src/basicTokenizer.cu (hunk 4 above)
[the remark this answers is not in the document]
Comment (brhodes10) [reply cpp/src/basicTokenizer.cu]: Removed transfer_cp_data_to_device function

**Comment by brhodes10** [comment]
> I just reviewed some of BasicTokenizer for now. Suggest making those changes and corresponding changes to the other tokenizers, and then I can re-review. This should simplify the changes a little.

Hi @harrism I have made changes based on your comments. I saw areas where I could apply your suggestions to the `fullTokenizer` as well. Let me know if you have any more comments. Thank you

File: cpp/CMakeLists.txt [hunk 6]
```diff
@@ -40,28 +223,27 @@ ADD_CUSTOM_TARGET(gen_cuda_header ALL
...
 
-target_link_libraries(tokenizer ${CU_TOK_LIBS})
+add_custom_target(build_benchmarks_clx
+                  DEPENDS ${BENCHMARK_LIST})
 
 add_dependencies(clx gen_cuda_header)
 
 install(TARGETS clx LIBRARY
-    DESTINATION lib)
+    DESTINATION lib)
```
Comment (harrism) [inline cpp/CMakeLists.txt:249]:
```suggestion
    DESTINATION lib)

```

**Comment by harrism** [comment]
I think something is wrong here?  124000 additions and deletions? Those files can't be that long, can they? 

image

**Comment by efajardo-nv** [comment]
rerun tests

**Pull request merged**